#PLOT PHS RESULTS 

#Plot results after Step2_PHS_output.R was ran.

rm(list=ls())

#INPUT FILES: 
SNP_annotations<-read.csv("Barley_Annotations.txt",header=T,sep="\t")
genmap<-read.table("GeneticMap_T3_020315",header=T)

#output generated by Step2_PHS_output.R
Allele_A_infile<-read.table("Analysis/Allele_A_all_16_pops.txt",header=T)
Allele_B_infile<-read.table("Analysis/Allele_B_all_16_pops.txt",header=T)

Allele_B_infile_temp<-Allele_B_infile
colnames(Allele_B_infile_temp)<-colnames(Allele_A_infile)

#combine significant allele A and B
All_significantPHS <-rbind(Allele_A_infile, Allele_B_infile_temp )

write.table(All_significantPHS,"All_significant_listPHS.txt",quote=F,row.names=F,col.names=T,sep="\t")

All_significantPHS<-read.table("/All_significant_listPHS.txt",header=T)


breeding_program<-as.data.frame(c("AB2","AB6","BA2","BA6","BAI2","MN6","MT2","N2","N6","OR2","OR6","UT2","UT6","VT6","WA2","WA6"))

#============Preparing Table S9. PHS significant
lengthA<-Allele_A_infile$Block_A2 - Allele_A_infile$Block_A1
Allele_A_infile_SIG<-Allele_A_infile[,-c(8,10:24)]
	Allele_A_infile_SIG<-cbind(Allele_A_infile_SIG,lengthA)
	colnames(Allele_A_infile_SIG)<-c("Population","Cumulative" ,"Chr", "SNP_index",  "SNP_name", "Genetic_dist" ,"PHS","Frequency","Length")

lengthB<-Allele_B_infile$Block_B2 - Allele_B_infile$Block_B1
Allele_B_infile_SIG<-Allele_B_infile[,-c(7,9,11:24)]
	Allele_B_infile_SIG<-cbind(Allele_B_infile_SIG,lengthB)
	colnames(Allele_B_infile_SIG)<-c("Population","Cumulative" ,"Chr", "SNP_index",  "SNP_name", "Genetic_dist" ,"PHS","Frequency","Length")

All_sig_phs<-rbind(Allele_A_infile_SIG, Allele_B_infile_SIG)


All_sig_phs_or<-All_sig_phs[order(All_sig_phs$Population, All_sig_phs$Chr,All_sig_phs$SNP_index),]
#remove extra columns Cumulative, SNP_index
All_sig_phs_or_rmv<-All_sig_phs_or[,-c(2,4)]
All_sig_phs_or_table<-All_sig_phs_or_rmv[,c(1,3,2,4,5,6,7)]

count_pops_phs<-as.data.frame(table(All_sig_phs_or_table[,1]))

#Construct table as for publication
for (p in 1:(dim(count_pops_phs)[1])){
	pop<-count_pops_phs[p,1]
	PHS_results_pop<-subset(All_sig_phs_or_table, All_sig_phs_or_table $Population == pop)
	
	#Remove pop assignation
	Pop_phs<-PHS_results_pop[,-1]
	#add a row with the population name
	
	POP_OR_PHS <-rbind(data.frame(SNP_name =as.factor(pop), Chr =1, Genetic_dist =1, PHS =1,Frequency=1,Length=1),Pop_phs)
	assign(paste(pop,"_PHS_sig",sep=""), POP_OR_PHS)
	
}
#combine PHS files for all populations
file_pops<-grep(x= ls(pos=1), pattern="_PHS_sig", value=TRUE)
TABLE_S_PHS_significant<-do.call(rbind,mget(file_pops))

TABLE_X11<-sub("X11_","11_", as.matrix(TABLE_S_PHS_significant))
TABLE_X12<-sub("X12_","12_", as.matrix(TABLE_X11))

TABLE_S_PHS_significant_ready<-as.data.frame(TABLE_X12)

write.table(TABLE_S_PHS_significant_ready,"TableS_PHS_sig_pops.txt",quote=F,row.names=F,col.names=T,sep="\t")


##==========TABLE S7=====

#Find the SNP annotations for those SNPs with significant PHS values
snp_annotation_phs<-SNP_annotations[(SNP_annotations$SNPName %in% All_significantPHS$SNP_name),]

#Order by SNP name. Then add chr# info and cM position
snp_list_or<-snp_annotation_phs[order(snp_annotation_phs[,1]),]

genmap_inSNPlist<-genmap[(genmap$SNP %in% snp_list_or[,1] ),]
genmap_inSNPlist_or<-genmap_inSNPlist[order(genmap_inSNPlist[,1]),]

if(identical(as.character(genmap_inSNPlist_or[,1]),as.character(snp_list_or[,1])) == FALSE) stop (print ('SNPs are in different order'))

list_w_positions<-cbind(genmap_inSNPlist_or, snp_list_or)
list_w_positions_orIndex<-list_w_positions[order(list_w_positions[,5]),]

list_w_positions_orIndex[,1] <-sub('X12_',"12_", list_w_positions_orIndex[,1])
list_w_positions_orIndex[,1] <-sub('X11_',"11_", list_w_positions_orIndex[,1])


write.table(list_w_positions_orIndex[,c(1,3,2,10,8)],"Annotations_significant_SNPs.txt",quote=F,row.names=F,col.names=T,sep="\t")



#Create an index list to order the samples in a more meaningfull way

Index_a <-as.character(Allele_A_infile[,1])
	Index_a[Index_a == 'AB2'] <- 'a'
	Index_a[Index_a == 'AB6'] <- 'b'
	Index_a[Index_a == 'BA2'] <- 'c'
	Index_a[Index_a == 'BA6'] <- 'd'
	Index_a[Index_a == 'BAI2'] <- 'e'
	Index_a[Index_a == 'MN6'] <- 'f'
	Index_a[Index_a == 'MT2'] <- 'g'
	Index_a[Index_a == 'N2'] <- 'h'
	Index_a[Index_a == 'N6'] <- 'i'
	Index_a[Index_a == 'OR2'] <- 'j'
	Index_a[Index_a == 'OR6'] <- 'k'
	Index_a[Index_a == 'UT2'] <- 'l'
	Index_a[Index_a == 'UT6'] <- 'm'
	Index_a[Index_a == 'VT6'] <- 'n'
	Index_a[Index_a == 'WA2'] <- 'o'
	Index_a[Index_a == 'WA6'] <- 'p'
	

Index_b <-as.character(Allele_B_infile[,1])
	Index_b[Index_b == 'AB2'] <- 'a'
	Index_b[Index_b == 'AB6'] <- 'b'
	Index_b[Index_b == 'BA2'] <- 'c'
	Index_b[Index_b == 'BA6'] <- 'd'
	Index_b[Index_b == 'BAI2'] <- 'e'
	Index_b[Index_b == 'MN6'] <- 'f'
	Index_b[Index_b == 'MT2'] <- 'g'
	Index_b[Index_b == 'N2'] <- 'h'
	Index_b[Index_b == 'N6'] <- 'i'
	Index_b[Index_b == 'OR2'] <- 'j'
	Index_b[Index_b == 'OR6'] <- 'k'
	Index_b[Index_b == 'UT2'] <- 'l'
	Index_b[Index_b == 'UT6'] <- 'm'
	Index_b[Index_b == 'VT6'] <- 'n'
	Index_b[Index_b == 'WA2'] <- 'o'
	Index_b[Index_b == 'WA6'] <- 'p'
	


Allele_A<-cbind(as.data.frame(Index_a), Allele_A_infile)
Allele_B<-cbind(as.data.frame(Index_b), Allele_B_infile)

breeding_program <- as.data.frame(c('a','b','c','d','e','f','g','h','i','j','k','l','m','n','o','p'))
#Separate file according to allele frequency. Looking at the summary.

summary(Allele_A)  ## Allele A can be in bins of 10% going up to 70%
summary(Allele_B)  ## Allele B going in bins of 10% going up to 60%

##Removed repeted alleles from Allele_B (arbitrary choice from where to remove)
no_rep_Allele_B<-subset(Allele_B,!(Allele_B$SNP_name %in% Allele_A$SNP_name))
## FOR ALLELE A

freq_10_19_A<-subset(Allele_A,Allele_A$Freq_A >=0.10 & Allele_A$Freq_A <= 0.20)
freq_20_29_A<-subset(Allele_A,Allele_A$Freq_A >0.20 & Allele_A$Freq_A <= 0.30)
freq_30_39_A<-subset(Allele_A,Allele_A$Freq_A >0.30 & Allele_A$Freq_A <= 0.40)
freq_40_49_A<-subset(Allele_A,Allele_A$Freq_A >0.40 & Allele_A$Freq_A <= 0.50)
freq_50_more_A<-subset(Allele_A,Allele_A$Freq_A >0.50 )

##FOR ALLELE B
freq_10_19_B<-subset(no_rep_Allele_B,no_rep_Allele_B$Freq_B >=0.10 & no_rep_Allele_B$Freq_B <= 0.20)
freq_20_29_B<-subset(no_rep_Allele_B,no_rep_Allele_B$Freq_B >0.20 & no_rep_Allele_B$Freq_B <= 0.30)
freq_30_39_B<-subset(no_rep_Allele_B,no_rep_Allele_B$Freq_B >0.30 & no_rep_Allele_B$Freq_B <= 0.40)
freq_40_49_B<-subset(no_rep_Allele_B,no_rep_Allele_B$Freq_B >0.40 & no_rep_Allele_B$Freq_B <= 0.50)
freq_50_more_B<-subset(no_rep_Allele_B,no_rep_Allele_B$Freq_B >0.50 )


##===== PLOT Figure S8 ==============

pdf("FigureS8.pdf",width=8,height=6)
par(mar=c(5.1, 11, 6.1, 12.1), xpd=TRUE) ## xpd indicates that we can go outside the ploting area
plot(Allele_A$ACCUMULATIVE,Allele_A$Index_a,xlab="", ylab="",cex=0.3, yaxt= "n",xaxt='n',ylim=c(0,17),xlim=c(0,1115), main=c("PHS in North American Breeding Programs"))

points(freq_10_19_A$ACCUMULATIVE,freq_10_19_A$Index_a,cex=0.5,col="purple",pch=16)
points(freq_20_29_A$ACCUMULATIVE,freq_20_29_A$Index_a,cex=0.5,col="blue",pch=16)
points(freq_30_39_A$ACCUMULATIVE,freq_30_39_A$Index_a,cex=0.5,col="green",pch=16)
points(freq_40_49_A$ACCUMULATIVE,freq_40_49_A$Index_a,cex=0.5,col="orange",pch=16)
points(freq_50_more_A$ACCUMULATIVE,freq_50_more_A$Index_a,cex=0.5,col="red",pch=16)


#plot(no_rep_Allele_B$ACCUMULATIVE,no_rep_Allele_B$Index_b,xlab="cM", ylab="",cex=0.5, yaxt= "n")
points(freq_10_19_B$ACCUMULATIVE,freq_10_19_B$Index_b,cex=0.5,col="purple",pch=16)
points(freq_20_29_B$ACCUMULATIVE,freq_20_29_B$Index_b,cex=0.5,col="blue",pch=16)
points(freq_30_39_B$ACCUMULATIVE,freq_30_39_B$Index_b,cex=0.5,col="green",pch=16)
points(freq_40_49_B$ACCUMULATIVE,freq_40_49_B$Index_b,cex=0.5,col="orange",pch=16)
points(freq_50_more_B$ACCUMULATIVE,freq_50_more_B$Index_b,cex=0.5,col="red",pch=16)



FREQUENCIES<-as.data.frame(c("freq_10_19_A","freq_20_29_A","freq_30_39_A","freq_40_49_A","freq_50_more_A"))

FREQUENCIES_B<-as.data.frame(c("freq_10_19_B","freq_20_29_B","freq_30_39_B","freq_40_49_B","freq_50_more_B"))


for(f in 1:(dim(FREQUENCIES)[1])) {

	freq_data<-get(paste(FREQUENCIES[f,1],sep=""))
	freq_data2<-get(paste(FREQUENCIES_B[f,1],sep=""))
	
	for (i in 1:(dim(breeding_program)[1])){

	pop_analyzed<-breeding_program[i,]
	POP_freq_A<-subset(freq_data , freq_data[,1] == as.vector(pop_analyzed))
	POP_freq_B<-subset(freq_data2 , freq_data2[,1] == as.vector(pop_analyzed))
	YAXIS<-rep(i,(dim(POP_freq_A)[1]))
	YAXIS_B<-rep(i,(dim(POP_freq_B)[1]))
	#NOISE<-c(0.09,0.12,0.15,0.18,0.21)
	COLOR<-c("purple","blue","green","orange","red")
	#segments(POP_freq_A $BlockA1_accum, YAXIS+(NOISE[f]), POP_freq_A $BlockA2_accum, YAXIS+(NOISE[f]),col=COLOR[f])
	segments(POP_freq_A$BlockA1_accum, YAXIS, POP_freq_A $BlockA2_accum, YAXIS,col=COLOR[f])
	segments(POP_freq_B$BlockB1_accum, YAXIS_B, POP_freq_B $BlockB2_accum, YAXIS_B,col=COLOR[f])
	}
}
	

#boundaries for chromosome
abline(v=c(142.6,316.14,496.27,642.76,832.66,974.87),xpd=FALSE, lty=3, col="black") 

#axis(2,at=1:16,labels=c("AB2","AB6","BA2","BA6","BAI2","MN6","MT2","N2","N6","OR2","OR6","UT2","UT6","VT6","WA2","WA6"),las=1)
axis(2,at=1:16,labels=c("Idaho (2)","Idaho (6)","Busch Ag.(2)","Busch Ag. (6)","Busch Ag. (Int.)","Minnesota (6)","Montana (2)","North Dakota (2)","North Dakota (6)","Oregon (2)","Oregon (6)","Utah (2)","Utah (6)","Virginia (6)","Washington (2)","Washington (6)"),las=1)

legend(inset=c(-0.4,0), cex=0.7, "bottomright",c("0.1-0.19","0.20-0.29","0.30-0.39","0.40-0.49",">=0.50", "SNP","Extent PHS"),col=c("purple","blue","green","orange","red","black","black"),pch=c(15,15,15,15,15,16,NA),lty=c(NA,NA,NA,NA,NA,NA,1),title=expression(bold("Allele Frequency")),title.adj=0.70)

axis(side=1,at=c(55.30,225.3,397.1,560.5, 738.2,900.6, 1058.0), labels=c("1H","2H","3H","4H","5H","6H","7H"))

dev.off()


